test-adapter/src/main/scala/org/scalajs/testing/adapter/HTMLRunnerBuilder.scala

```diff
@@ -26,85 +27,87 @@ import org.scalajs.testing.common._
 
 /** Template for the HTML runner. */
 object HTMLRunnerBuilder {
-
-  private val tmpSuffixRE = """[a-zA-Z0-9-_.]*$""".r
-
-  private def tmpFile(path: String, in: InputStream): URI = {
-    try {
-      /* - createTempFile requires a prefix of at least 3 chars
-       * - we use a safe part of the path as suffix so the extension stays (some
-       *   browsers need that) and there is a clue which file it came from.
-       */
-      val suffix = tmpSuffixRE.findFirstIn(path).orNull
-
-      val f = File.createTempFile("tmp-", suffix)
-      f.deleteOnExit()
-      Files.copy(in, f.toPath(), StandardCopyOption.REPLACE_EXISTING)
-      f.toURI()
-    } finally {
-      in.close()
+  @deprecated("Use write instead", "1.2.0")
+  def writeToFile(output: File, title: String, input: Seq[Input],
+      frameworkImplClassNames: List[List[String]],
+      taskDefs: List[TaskDef]): Unit = {
+    val outputPath = output.toPath()
+    val artifactsDir =
+      Files.createTempDirectory(outputPath.getParent(), ".html-artifacts")
+
+    sys.addShutdownHook {
+      Files.walkFileTree(artifactsDir, new SimpleFileVisitor[Path] {
+        override def visitFile(file: Path, attrs: BasicFileAttributes): FileVisitResult = {
+          Files.delete(file)
+          FileVisitResult.CONTINUE
+        }
+
+        override def postVisitDirectory(dir: Path, exc: IOException): FileVisitResult = {
+          Files.delete(dir)
+          FileVisitResult.CONTINUE
+        }
+      })
     }
+
+    write(outputPath, artifactsDir, title, input, frameworkImplClassNames, taskDefs)
   }
 
-  def writeToFile(output: File, title: String, input: Seq[Input],
+  def write(output: Path, artifactsDir: Path, title: String, input: Seq[Input],
       frameworkImplClassNames: List[List[String]],
       taskDefs: List[TaskDef]): Unit = {
 
-    val jsFiles = input.map {
-      case Input.Script(script) => script
+    def artifactPath(name: String): (String, Path) = {
+      val path = artifactsDir.resolve(name)
+      (output.getParent().relativize(path).toString(), path)
```

Comment:
On Windows, this produces `\`s in the relativized path, which are not really valid in HTML. We can avoid this issue by relativize the paths' respective URIs:
```suggestion
      (output.getParent().toUri.relativize(path.toUri).toString(), path)
```

Reply:
Ugh... Right. That approach has two problems:

- It doesn't fail if relativization doesn't work (`Path#relativize` OTOH does).
- It doesn't relativize to the parent directory (which is likely the primary use case if somebody changed `artifactPath in testHtml` manually).

I couldn't figure out a way to fix this on the quick :-/ Need to think more.

Reply:
I you want to keep using `Path.relativize`, then you can simply do `.replace('\\', '/')`.

Reply:
Updated.